<launch>
    <arg name="x_goal" default="2.25" />
    <arg name="y_goal" default="-1" />

    <arg name="world_frame" default="map" />
    <arg name="robot_frame" default="base_link" />
    <arg name="cmd_vel_topic" default="/cmd_vel" />
    <arg name="local_map_topic" default="/local_map" />
    <arg name="laser_topic" default="/laser/scan" />
    <arg name="path_topic" default="/rrt_path" />
    <arg name="control_path_topic" default="local_path" />

    <arg name="robot_density" default="100" />
    <arg name="robot_height" default="0.35" />
    <arg name="robot_width" default="0.23" />

    <arg name="run_rrt_star" default="false" />

    <arg name="map_topic_for_local_rrt" default="/map_topic_for_local_rrt" />
    <arg name="map_frame_for_local_rrt" default="map_frame_for_local_rrt" />

    <node pkg="rrt_project" type="emergency_button.py" name="emergency_button" output="screen" required="true"/>

    <node pkg="rrt_project" type="map_server.py" name="map_server_node"/>
    <node pkg="rrt_project" type="merge_local_map" name="merge_local_map" output="screen"/>

    <node pkg="rrt_project" type="global_rrt" name="global_rrt" output="screen" unless="$(arg run_rrt_star)">
        <rosparam file="$(find rrt_project)/config/limo_rrt.yaml" command="load" />
        <param name="robot_frame" value="$(arg robot_frame)" />
        <param name="world_frame" value="$(arg world_frame)" />
        <param name="path_topic" value="$(arg path_topic)" />
        <param name="x_goal" value="$(arg x_goal)" />
        <param name="y_goal" value="$(arg y_goal)" />
        <param name="rate" value="1.0" />
    </node>

    <node pkg="rrt_project" type="global_rrt_star" name="global_rrt_star" output="screen" if="$(arg run_rrt_star)">
        <rosparam file="$(find rrt_project)/config/limo_rrt.yaml" command="load" />
        <param name="robot_frame" value="$(arg robot_frame)" />
        <param name="world_frame" value="$(arg world_frame)" />
        <param name="path_topic" value="$(arg path_topic)" />
        <param name="x_goal" value="$(arg x_goal)" />
        <param name="y_goal" value="$(arg y_goal)" />
        <param name="rate" value="1.0" />
    </node>

    <include file="$(find rrt_project)/launch/launches_for_dissertation/simulation/mapping/limo_mapping.launch">
        <arg name="scan_topic" value="$(arg laser_topic)" />
        <arg name="map_topic" value="$(arg map_topic_for_local_rrt)" />
        <arg name="map_frame" value="$(arg map_frame_for_local_rrt)" />
        <arg name="base_frame" value="$(arg robot_frame)" />
        <arg name="resolution" value="0.01" />
        <arg name="x_min" value="-4.0" />
        <arg name="y_min" value="-2.5" />
        <arg name="x_max" value="4.0" />
        <arg name="y_max" value="2.5" />
    </include>

    <!-- Max rate is over 1000hz with this specs -->
    <node pkg="rrt_project" type="local_map" name="local_map">
        <param name="global_map_topic" value="$(arg map_topic_for_local_rrt)" />
        <param name="local_map_topic" value="$(arg local_map_topic)" />
        <param name="pose_topic" value="/vrpn_client_node/P1/pose" />
        <param name="width_meters" value="1.5" />
        <param name="height_meters" value="1.5" />
        <param name="publish_rate" value="10.0" />
        <param name="local_resolution" value="0.01" />
    </node>

    <!-- Max rate is 230hz with this specs -->
    <node pkg="rrt_project" type="local_rrt" name="local_rrt" output="screen"> 
        <rosparam file="$(find rrt_project)/config/limo_local_rrt.yaml" command="load" />
        <param name="robot_frame" value="$(arg robot_frame)" />
        <param name="step_size" value="7.0" />
        <param name="local_map_topic" value="$(arg local_map_topic)" />
        <param name="path_topic" value="$(arg path_topic)" />
        <param name="rate" value="100.0" />
    </node>

    <!-- Control and potential nodes -->
    <node pkg="rrt_project" type="differential_robot_control_path.py" name="limo_control" output="screen">
        <rosparam file="$(find rrt_project)/config/simulation/differential_robots/pioneer.yaml" command="load" />
        <param name="path_topic" value="$(arg control_path_topic)" />
        <param name="robot_frame" value="$(arg robot_frame)" />
        <param name="world_frame" value="$(arg world_frame)" />
        <param name="robot_control_topic" value="$(arg cmd_vel_topic)"/>
    </node>

    <node pkg="rrt_project" type="obstacle_avoidance" name="obstacle_avoidance_node" output="screen">
        <param name="num_clusters" value="12"/>
        <param name="use_angle_filter" value="true"/>
        <param name="angle_filter_margin" value="0.2"/>

        <param name="filter_type" value="0"/>
        <param name="filter_gain" value="0.8"/>

        <param name="robot_height" value="$(arg robot_height)"/>
        <param name="robot_width" value="$(arg robot_width)"/>
        <param name="x_offset" value="0.1"/>
        <param name="lidar_topic" value="$(arg laser_topic)"/>
        
        <param name="min_dist_threshold" value="0.3"/>

        <param name="obstacle_avoidance_n" value="2.0"/>
        <param name="obstacle_avoidance_a" value="0.3"/>
        <param name="obstacle_avoidance_b" value="0.3"/>
        <param name="obstacle_avoidance_k" value="0.3"/>
        <param name="obstacle_avoidance_lambda" value="0.001"/>
        <param name="potential_gain" value="1.1"/>

        <param name="saturate_potential" value="100.0"/>
    </node>

    <node name="traveled_path" pkg="rrt_project" type="publish_robot_traveled_path.py" output="screen">
        <!-- <param name="pose_topic" value="/robot_pose"/>
        <param name="path_topic" value="/robot_path"/> -->
        <param name="rate" value="10"/>
    </node>
</launch>